### ISSUE TITLE
Registry cli with secured calls (https) and no authentication

### ISSUE THREAD

**Issue by aerri** [apigee/registry#853]
I have a Registry server (development instance) deployed on my organization's infrastructure, so, the endpoint uses https protocol. When using the registry cli with the property registry.insecure set to false, it makes usage of the google authentication process which can not be bypassed.

**Comment by theganyo** [comment]
Hi, @aerri. I'm investigating your issue and need some clarification: Could you explain what you mean by "makes usage of the google authentication process which can not be bypassed"? Thanks!

**Comment by aerri** [comment]
I have a Registry Server deployed in the infrastructure of the company I work, so I use the CLI with the insecure flag marked as false, as you can see on the picture below.

![image](https://user-images.githubusercontent.com/26742294/204519897-4e82c0b2-e799-41da-84ba-f527eaf7d488.png)

When I send commands, I get an error like this on the picture below.

![image](https://user-images.githubusercontent.com/26742294/204520067-180c7c55-092a-4f74-8474-8262f7c083c8.png)
![image](https://user-images.githubusercontent.com/26742294/204520161-57c02921-3555-450a-bb40-3e874f5b73ec.png)

**Comment by theganyo** [comment]
@aerri - The error message you posted indicates that the URL you're connecting to is `http` and it sounds like you want `https`. The default value for `insecure` is false, but if there is an env var `APG_REGISTRY_INSECURE` with any value, that could be interpreted as a true value - so if you have that, please try to `unset` that value.

**Comment by aerri** [comment]
> @aerri - The error message you posted indicates that the URL you're connecting to is `http` and it sounds like you want `https`. The default value for `insecure` is false, but if there is an env var `APG_REGISTRY_INSECURE` with any value, that could be interpreted as a true value - so if you have that, please try to `unset` that value.

I'm already using with insecure set to false. I'm using along with registry config command.

**Comment by theganyo** [comment]
Ok, thanks. I just wanted to ensure that there wasn't an env var that might be interfering - and also to let you know that you don't need to specify this flag.

**Comment by theganyo** [comment]
@aerri , Could you be seeing an issue with the CLI attempting to retrieve a token for. you? Run `registry auth -h` for more information on how that works.

**Comment by timburks** [comment]
Experimenting to try to reproduce this, I deployed an instance to Cloud Run and did not include `allUsers` in the `Cloud Run Invoker` role.  After clearing my token source (below), calls from the registry tool failed with the message below:
```
$ registry config set token-source ""
Updated property "token-source".
$ registry list projects
 FATAL[0000] Failed to match or handle command error=rpc error: code = Unauthenticated desc = unexpected HTTP status code received from server: 401 (Unauthorized); transport: received unexpected content-type "text/html; charset=UTF-8" uid=1b08ca42
```

**Comment by aerri** [comment]
> @aerri , Could you be seeing an issue with the CLI attempting to retrieve a token for. you? Run `registry auth -h` for more information on how that works.

I tried doing this, even when I explicitly set the token to an empty string, it tries to run that field as a command. The deployment that we doesn't need any authentication process, but the CLI does. I was wondering if this authentication could be opt out.

**Comment by aerri** [comment]
> Experimenting to try to reproduce this, I deployed an instance to Cloud Run and did not include `allUsers` in the `Cloud Run Invoker` role.  After clearing my token source (below), calls from the registry tool failed with the message below:
> ```
> $ registry config set token-source ""
> Updated property "token-source".
> $ registry list projects
>  FATAL[0000] Failed to match or handle command error=rpc error: code = Unauthenticated desc = unexpected HTTP status code received from server: 401 (Unauthorized); transport: received unexpected content-type "text/html; charset=UTF-8" uid=1b08ca42
> ``` 
> After I re-added `allUsers` to `Cloud Run Invoker`, requests succeeded (after a short propagation time).
> 
> tl;dr @aerri I'm wondering if your backend is expecting some authentication that isn't being sent.

No, we just deployed the server without any authentication. We're not using the GKE deployment system or something like that. So, it has no dependency with GCloud services.

**Comment by aerri** [comment]
I'll try yo debug the code tomorrow and trace all the steps the program makes so I can explain it better to you, guys.

**Comment by theganyo** [comment]
> even when I explicitly set the token to an empty string, it tries to run that field as a command

I'm unsure how this would be possible, as this is the code: https://github.com/apigee/registry/blob/main/pkg/config/configuration.go#L122

But you could also try using `--registry.token ""` to perhaps see if you can bypass this.

**Comment by aerri** [comment]
> > even when I explicitly set the token to an empty string, it tries to run that field as a command
> 
> I'm unsure how this would be possible, as this is the code: https://github.com/apigee/registry/blob/main/pkg/config/configuration.go#L122
> 
> But you could also try using `--registry.token ""` to perhaps see if you can bypass this.

Already tried. haha I'm gonna try it once again and bring back the results to you.

**Comment by aerri** [comment]
I think I've found the cause. When config.Insecure is true, the CLI calls the grpc.Dial with the options from the config, otherwise, it will use the configurations from a method called defaultAdminGRPCClientOptions(). It always makes usage of the endpoints from google, I guess.

![image](https://user-images.githubusercontent.com/26742294/204812590-52d77754-6388-4f00-8c5a-34876b7ddc00.png)
![image](https://user-images.githubusercontent.com/26742294/204812766-1f8415a1-7e48-458d-b272-a6fbcfe0edce.png)
![image](https://user-images.githubusercontent.com/26742294/204812845-70ec70bb-cfad-47cd-8170-2826a20a72ff.png)

**Comment by timburks** [comment]
@aerri One puzzling thing about that hypothesis is that the line above `if config.Insecure {` is [this one](https://github.com/apigee/registry/blob/8ead1d0431e20db00ab16baa7acfc5589e29498a/pkg/connection/client.go#L33), which configures the connection to use `registry.address` from your configuration:
```
	opts = append(opts, option.WithEndpoint(config.Address))
```

To test your hypothesis locally, if you're building the `registry` tool from source, could you delete the `internaloption.WithDefaultEndpoint()` and two subsequent calls from `defaultAdminGRPCClientOptions()` and `defaultRegistryGRPClientOptions()` ? (i.e. lines 57-59 [here](https://github.com/apigee/registry/blob/223bb5255a9d62a3296bec0e4ca2f55d23c7bc86/gapic/admin_client.go#L57) and lines 82-84 [here](https://github.com/apigee/registry/blob/223bb5255a9d62a3296bec0e4ca2f55d23c7bc86/gapic/registry_client.go#L82)) and try your updated build?

When I did that, I was able to run the `registry` tool with a variety of local and remote services with no noticeable effects.

Also, could you include the results of running `registry config list` followed by a call to your backend with the full error message that is returned?  Something like the following (which is working for me, hence no error), and please obscure the `registry.address` (as I've done also).
```
$ registry config list
registry.address = registry-XXXXXXXXXX-YY.a.run.app:443
registry.insecure = false
registry.location = global
registry.project = unbundled

Your active configuration is: "run".
$ registry rpc admin get-project --name projects/test --json
{
  "name": "projects/test",
  "createTime": "2022-11-16T21:49:05.322456Z",
  "updateTime": "2022-11-16T21:49:05.322456Z"
}
```

**Comment by aerri** [comment]
Sure, I'll try that.

**Comment by aerri** [comment]
> @aerri One puzzling thing about that hypothesis is that the line above `if config.Insecure {` is [this one](https://github.com/apigee/registry/blob/8ead1d0431e20db00ab16baa7acfc5589e29498a/pkg/connection/client.go#L33), which configures the connection to use `registry.address` from your configuration:
> 
> ```
> 	opts = append(opts, option.WithEndpoint(config.Address))
> ```
> 
> To test your hypothesis locally, if you're building the `registry` tool from source, could you delete the `internaloption.WithDefaultEndpoint()` and two subsequent calls from `defaultAdminGRPCClientOptions()` and `defaultRegistryGRPClientOptions()` ? (i.e. lines 57-59 [here](https://github.com/apigee/registry/blob/223bb5255a9d62a3296bec0e4ca2f55d23c7bc86/gapic/admin_client.go#L57) and lines 82-84 [here](https://github.com/apigee/registry/blob/223bb5255a9d62a3296bec0e4ca2f55d23c7bc86/gapic/registry_client.go#L82)) and try your updated build?
> 
> When I did that, I was able to run the `registry` tool with a variety of local and remote services with no noticeable effects.
> 
> Also, could you include the results of running `registry config list` followed by a call to your backend with the full error message that is returned? Something like the following (which is working for me, hence no error), and please obscure the `registry.address` (as I've done also).
> 
> ```
> $ registry config list
> registry.address = registry-XXXXXXXXXX-YY.a.run.app:443
> registry.insecure = false
> registry.location = global
> registry.project = unbundled
> 
> Your active configuration is: "run".
> $ registry rpc admin get-project --name projects/test --json
> {
>   "name": "projects/test",
>   "createTime": "2022-11-16T21:49:05.322456Z",
>   "updateTime": "2022-11-16T21:49:05.322456Z"
> }
> ```

I did that and received this response:

![image](https://user-images.githubusercontent.com/26742294/205042341-a2443902-6a2f-4898-9b14-0493a15be899.png)

**Comment by aerri** [comment]
I installed the gcloud cli and set it up the authentication and now it worked. I didn't know that in order to use the registry tool I needed to set up the gcloud cli, even if the Registry server it's not deployed in Google's infrastructure.

**Comment by theganyo** [comment]
Excellent news! Thank you for that information, I can now recreate your issue.

**Comment by theganyo** [comment]
@aerri Ok. I've found the issue: When the config specifies "insecure: false", security constraints require some kind of credentials. If there isn't a token specified in the config or a flag, the system attempts to get the default gcloud token for you. The error(s) you saw were related to that. I'll improve the error message and update our docs so this is more clear.

The solution is to specify a token in the config. If your server doesn't do any kind of authorization, the token can be anything you wish, but the token is required to be provided if there isn't a default token.

**Comment by aerri** [comment]
> @aerri Ok. I've found the issue: When the config specifies "insecure: false", security constraints require some kind of credentials. If there isn't a token specified in the config or as a flag, the system attempts to get the default gcloud token for you. The error(s) you saw were related to that. I'll improve the error message and update our docs so this is more clear.
> 
> The solution is to specify a token in the config. If your server doesn't do any kind of authorization, the token can be anything you wish, but the token is required to be provided if there isn't a default token.

I see. I believe that the documentation mentions that the gcloud cli is a requirement, but I thought it was required only when using the registry tool to access the deployment hosted on Google Cloud. Once again, thank you for the attention and help.

**Comment by theganyo** [comment]
I've also created a new wiki page on configuration: https://github.com/apigee/registry/wiki/Registry-Tool-Configuration. Please let me know if anything is unclear or can be improved. Thanks!
